pkg/controller/controlplane/valuesprovider.go

```diff
@@ -303,13 +333,79 @@ func isOverlayEnabled(networking *gardencorev1beta1.Networking) (bool, error) {
 }
 
 // getControlPlaneShootChartValues collects and returns the control plane shoot chart values.
-func (vp *valuesProvider) getControlPlaneShootChartValues(cluster *extensionscontroller.Cluster) (map[string]interface{}, error) {
+func (vp *valuesProvider) getControlPlaneShootChartValues(cluster *extensionscontroller.Cluster, cp *apismetal.ControlPlaneConfig) (map[string]interface{}, error) {
 	if cluster.Shoot == nil {
 		return nil, fmt.Errorf("cluster %s does not contain a shoot object", cluster.ObjectMeta.Name)
 	}
 
+	metallb, err := getMetallbChartValues(cp)
+	if err != nil {
+		return nil, err
+	}
 	return map[string]interface{}{
 		metal.CloudControllerManagerName: map[string]interface{}{"enabled": true},
+		metal.MetallbName:                metallb,
 	}, nil
+}
 
+// getMetallbChartValues collects and returns the CCM chart values.
+func getMetallbChartValues(
+	cpConfig *apismetal.ControlPlaneConfig,
+) (map[string]interface{}, error) {
+	if cpConfig.LoadBalancerConfig == nil || cpConfig.LoadBalancerConfig.MetallbConfig == nil {
+		return map[string]interface{}{
+			"enabled": false,
+		}, nil
+	}
+
+	for _, cidr := range cpConfig.LoadBalancerConfig.MetallbConfig.IPAddressPool {
+		if err := parseAddressPool(cidr); err != nil {
+			return nil, fmt.Errorf("invalid CIDR %q in pool: %w", cidr, err)
+		}
+	}
+
+	enableSpeaker := false
+	if cpConfig.LoadBalancerConfig.MetallbConfig.EnableSpeaker {
+		enableSpeaker = true
+	}
+	enableL2 := false
+	if cpConfig.LoadBalancerConfig.MetallbConfig.EnableL2Advertisement {
+		enableL2 = true
+	}
+	return map[string]interface{}{
+		"enabled": true,
+		"speaker": map[string]interface{}{
+			"enabled": enableSpeaker,
+		},
+		"l2Advertisement": map[string]interface{}{
+			"enabled": enableL2,
+		},
+		"ipAddressPool": cpConfig.LoadBalancerConfig.MetallbConfig.IPAddressPool,
+	}, nil
+}
+
+func parseAddressPool(cidr string) error {
+	if !strings.Contains(cidr, "-") {
+		_, _, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return fmt.Errorf("invalid CIDR %q", cidr)
+		}
+		return nil
+	}
+	fs := strings.SplitN(cidr, "-", 2)
+	if len(fs) != 2 {
+		return fmt.Errorf("invalid IP range %q", cidr)
+	}
+	start := net.ParseIP(strings.TrimSpace(fs[0]))
+	if start == nil {
+		return fmt.Errorf("invalid IP range %q: invalid start IP %q", cidr, fs[0])
+	}
+	end := net.ParseIP(strings.TrimSpace(fs[1]))
+	if end == nil {
+		return fmt.Errorf("invalid IP range %q: invalid end IP %q", cidr, fs[1])
+	}
+	if bytes.Compare(start, end) > 0 {
```

Comment:
If both IPs are invalid `start` and `end` will be `nil`. `Compare` returns zero in that case, passing the current validation logic.